src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp

```diff
@@ -0,0 +1,175 @@
+#include <cmath>
+#include <cstring>
+#include <limits>
+#include <pmmintrin.h>
+#include "../Stdafx.h"
+
+// Compute the output value of the field-aware factorization, as the sum of the linear part and the latent part. 
+// The linear part is the inner product of linearWeights and featureValues.
+// The latent part is the sum of all intra-field interactions in one field f, for all fields possible. 
+EXPORT_API(void) CalculateIntermediateVariablesNative(int fieldCount, int latentDim, int count, _In_ int * fieldIndices, _In_ int * featureIndices, _In_ float * featureValues,
+    _In_ float * linearWeights, _In_ float * latentWeights, _Inout_ float * latentSum, _Out_ float * response)
+{
+    // The number of all possible fields
+    const int m = fieldCount;
+    const int d = latentDim;
+    const int c = count;
+    const int * pf = fieldIndices;
+    const int * pi = featureIndices;
+    const float * px = featureValues;
+    const float * pw = linearWeights;
+    const float * pv = latentWeights;
+    float * pq = latentSum;
+    float linearResponse = 0;
+    float latentResponse = 0;
+
+    memset(pq, 0, sizeof(float) * m * m * d);
+    __m128 _y = _mm_setzero_ps();
+    __m128 _tmp = _mm_setzero_ps();
+
+    for (int i = 0; i < c; i++)
+    {
+        const int f = pf[i];
+        const int j = pi[i];
+        linearResponse += pw[j] * px[i];
+
+        const __m128 _x = _mm_load1_ps(px + i);
+        const __m128 _xx = _mm_mul_ps(_x, _x);
+
+        // tmp -= <v_j,f, v_j,f> * x * x
+        const int vBias = j * m * d + f * d;
+
+        // j-th feature's latent vector in the f-th field hidden space.
+        const float * vjf = pv + vBias;
+
+        for (int k = 0; k + 4 <= d; k += 4)
+        {
+            const __m128 _v = _mm_load_ps(vjf + k);
+            _tmp = _mm_sub_ps(_tmp, _mm_mul_ps(_mm_mul_ps(_v, _v), _xx));
+        }
+
+        for (int fprime = 0; fprime < m; fprime++)
+        {
+            const int vBias = j * m * d + fprime * d;
```

Comment:
Overall I see many repeated computations in the tight loop.  Not sure if compiler will optimize them all out, but it seems like it might be worth at least trying to see if this refactored - both for readibilily and potential speed gains.

Reply:
Since this is orthogonal to porting, I'll log an issue and defer optimizing till after the tests get enabled. 

---
In reply to: [197932033](https://github.com/dotnet/machinelearning/pull/383#discussion_r197932033) [](ancestors = 197932033)

Reply:
@glebuk , we have discussed this before. According to my experience when designing LIBFFM, complier is not able to do so even if we finely tuned the complier configuration. We have also concluded that it's very difficult to refactorize them. Maybe some of them are similar, but you will see that a refactorization easily lead to extra loops, temporal variables, or function pointer. Again, please don't expect readability of SSE and training algorithm. I personally prefer to use document as a solution. Honestly, although our SSE math library has some sort of inline documentation, they are still not readable due to the lack of math formulations to me.

Reply:
I made several innocuous seeming changes for pre/post increment on loops and optimization of tight loops through intermediate calculations. 

Run through the test to make sure no regression, and the test run is aborted with an empty string as error message. 

Will log an "enhancement" about the optimizations, and take that on post enabling the test. 

---
In reply to: [197932033](https://github.com/dotnet/machinelearning/pull/383#discussion_r197932033) [](ancestors = 197932033)